feat: highlight search term in labels in dropdown list

BenjaminLefevreGithub opened the pull request:
https://github.com/ng-select/ng-select/issues/152

highlight search term in the items with a default style 'hightlight' (bold and underline text)

anjmao commented:
@BenjaminLefevreGithub Thanks for contributing. Could you please add some unit tests under Filter section https://github.com/ng-select/ng-select/blob/master/src/ng-select/ng-select.component.spec.ts#L976

anjmao commented on src/ng-select/ng-select.component.html:
```diff
@@ -53,7 +53,7 @@
 
             <ng-template #defaultOptionTemplate>
-                <span class="ng-option-label" [innerHTML]="item.label"></span>
+                <span class="ng-option-label" [innerHTML]="item.highlightedLabel"></span>
```
This approach is limited to default filter only and will not work if using custom [typeahead]. It would be better to create directive [ngOptionHighlight]="filterValue" which would be responsible for highlighting label. 

```html
<span [ngOptionHighlight]="filterValue" class="ng-option-label" [innerHTML]="item.label"></span>
```

Also users will be able to apply this directive even if they are using custom option templates.

BenjaminLefevreGithub replied:
Yep good idea, done.

anjmao commented on src/ng-select/items-list.ts:
```diff
@@ -179,12 +182,29 @@ export class ItemsList {
     }
 
+    private _highlightLabelWithSearchTerm(option: NgOption, term: string, indexOfTerm: number) {
+        if (indexOfTerm > -1 && !option.selected && !option.disabled) {
```
It seems to add unneeded characters in some cases.
![image](https://user-images.githubusercontent.com/3028012/35777365-60fb9614-09b5-11e8-8024-a28895945922.png)

BenjaminLefevreGithub replied:
fixed the end index was wrong on the third substring

BenjaminLefevreGithub commented:
@anjmao  thanks for the review. 

I fixed the bug (mistake on substring ;) the end index is not included :p) 

And tried to make what you said: make a directive. 
I updated also the template demos to take into account the directive. 

I fixed also a bug discovered during my tests (https://ng-select.github.io/ng-select#/data-sources : console error log when trying to filter on the object array (true, 'Two', 3).

anjmao commented on src/ng-select/ng-option-highlight.directive.ts:
```diff
@@ -0,0 +1,42 @@
+@Directive({
+    selector: '[ngOptionHighlight]'
+})
+export class NgOptionHighlightDirective implements OnChanges {
```
Looks nice, only add some unit tests for this directive and I can merge it. You can create new test file called `ng-option-highlight.directive.spec.ts`.

BenjaminLefevreGithub replied:
No problem I will add some unit tests tonight

BenjaminLefevreGithub commented:
No problem I will add some unit tests tonight

BenjaminLefevreGithub commented:
Just add unit test for the directive
